### PR TITLE
IGNITE-17534 Introduce HashIndexStorage interface and test implementation

### PR DESCRIPTION
**Pull request by sashapolo**
https://issues.apache.org/jira/browse/IGNITE-17534

### CODE REVIEW

File: modules/storage-api/src/main/java/org/apache/ignite/internal/storage/engine/MvTableStorage.java [hunk 1]
```diff
@@ -78,6 +79,16 @@ public interface MvTableStorage {
...
 
+    /**
+     * Returns an already created Hash Index with the given name or {@code null} if it does not exist.
+     *
+     * @param partitionId Partition ID for which this index has been built.
+     * @param indexName Index name.
+     * @return Hash Index storage or {@code null} if it does not exist..
+     */
+    @Nullable
+    HashIndexStorage getHashIndex(int partitionId, String indexName);
```
Comment (tkalkirill) [inline modules/storage-api/src/main/java/org/apache/ignite/internal/storage/engine/MvTableStorage.java:90]:
```suggestion
    /**
     * Returns an already created Hash Index with the given name or {@code null} if it does not exist.
     *
     * @param partitionId Partition ID for which this index has been built.
     * @param indexName Index name.
     * @return Hash Index storage or {@code null} if it does not exist.
     */
    @Nullable HashIndexStorage getHashIndex(int partitionId, String indexName);
```

Comment (tkalkirill) [inline modules/storage-api/src/main/java/org/apache/ignite/internal/storage/engine/MvTableStorage.java:90]: Maybe **TableIndexConfiguration** instead **indexName**?

File: modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/AbstractPageMemoryTableStorage.java [hunk 2]
```diff
@@ -151,6 +152,12 @@ public void createIndex(String indexName) {
         throw new UnsupportedOperationException("Not implemented yet");
     }
 
+    @Override
+    @Nullable
+    public HashIndexStorage getHashIndex(int partitionId, String indexName) {
+        throw new UnsupportedOperationException("Not implemented yet");
+    }
+
```
Comment (tkalkirill) [inline modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/AbstractPageMemoryTableStorage.java:160]:
```suggestion
    @Override
    public @Nullable HashIndexStorage getHashIndex(int partitionId, String indexName) {
        throw new UnsupportedOperationException("Not implemented yet");
    }

```

File: modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/RocksDbTableStorage.java [hunk 3]
```diff
@@ -426,6 +427,12 @@ public void createIndex(String indexName) {
         throw new UnsupportedOperationException("Not implemented yet");
     }
 
+    @Override
+    @Nullable
+    public HashIndexStorage getHashIndex(int partitionId, String indexName) {
+        throw new UnsupportedOperationException("Not implemented yet");
+    }
```
Comment (tkalkirill) [inline modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/RocksDbTableStorage.java:434]:
```suggestion
    @Override
    public @Nullable HashIndexStorage getHashIndex(int partitionId, String indexName) {
        throw new UnsupportedOperationException("Not implemented yet");
    }
```

File: modules/storage-api/src/main/java/org/apache/ignite/internal/storage/engine/MvTableStorage.java (hunk 1 above)
Comment (sashapolo) [reply modules/storage-api/src/main/java/org/apache/ignite/internal/storage/engine/MvTableStorage.java]: Why?

Comment (sashapolo) [reply modules/storage-api/src/main/java/org/apache/ignite/internal/storage/engine/MvTableStorage.java]: I personally don't like this formatting, sorry

File: modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/AbstractPageMemoryTableStorage.java (hunk 2 above)
Comment (sashapolo) [reply modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/AbstractPageMemoryTableStorage.java:160]: Do we have such formatting in the code style? If no, I would prefer to leave it as-is

File: modules/storage-api/src/test/java/org/apache/ignite/internal/storage/index/impl/TestHashIndexStorage.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    private final ConcurrentMap<ByteBuffer, Set<RowId>> index;
+
+    private final HashIndexDescriptor descriptor;
+
+    /**
+     * Constructor.
+     */
+    public TestHashIndexStorage(HashIndexDescriptor descriptor) {
+        this.descriptor = descriptor;
+        this.index = new ConcurrentHashMap<>();
```
Comment (rpuch) [inline modules/storage-api/src/test/java/org/apache/ignite/internal/storage/index/impl/TestHashIndexStorage.java:47]: Could we initialize this on the field declaration?

File: modules/storage-api/src/main/java/org/apache/ignite/internal/storage/index/HashIndexStorage.java [hunk 5]
```diff
@@ -17,21 +17,38 @@
...
 import org.apache.ignite.internal.schema.BinaryTuple;
 import org.apache.ignite.internal.storage.RowId;
 
 /**
- * Temporary API for creating Index rows from a list of column values. All columns must be sorted according to the index columns order,
- * specified by the {@link SortedIndexDescriptor#indexColumns()}.
+ * Storage for a Hash Index.
+ *
+ * <p>This storage serves as an unordered mapping from a subset of a table's columns (a.k.a. index columns) to a set of {@link RowId}s
+ * from a {@link org.apache.ignite.internal.storage.MvPartitionStorage} from the same table.
```
[the remark this answers is not in the document]
Comment (sashapolo) [reply modules/storage-api/src/main/java/org/apache/ignite/internal/storage/index/HashIndexStorage.java:28]: All indices store data from a single partition. I will clarify

File: modules/storage-api/src/test/java/org/apache/ignite/internal/storage/index/AbstractHashIndexStorageTest.java [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+    ) TableConfiguration tableCfg) {
+        createTestTable(tableCfg);
+
+        this.tableCfg = tableCfg;
+        this.indexStorage = createIndexStorage();
+        this.serializer = new BinaryTupleRowSerializer(indexStorage.indexDescriptor());
+    }
+
+    /**
+     * Configures a test table with columns of all supported types.
```
[the remark this answers is not in the document]
Comment (sashapolo) [reply modules/storage-api/src/test/java/org/apache/ignite/internal/storage/index/AbstractHashIndexStorageTest.java:90]: no. Do you suggest to create an index with all supported types? I'm not sure it's worth the complexity if we are talking about hash indices as they don't actually care about what's inside a BinaryTuple

[the remark this answers is not in the document]
Comment (rpuch) [reply modules/storage-api/src/test/java/org/apache/ignite/internal/storage/index/AbstractHashIndexStorageTest.java:90]: I mean, the comment should match the code. If the code is fine, let's change the comment.

[the remark this answers is not in the document]
Comment (sashapolo) [reply modules/storage-api/src/test/java/org/apache/ignite/internal/storage/index/AbstractHashIndexStorageTest.java:90]: oh, that's a copy-paste mistake, will fix it, thank you

**Comment by rpuch** [comment]
Tests have failed. This can be fixed by adding `UnlimitedBudgetConfigurationSchema.class` to the list of the polymorphic configurations.

**Review by rpuch — APPROVED** [review]

**Pull request merged**